internal/auth/oauth_client.go

```diff
@@ -0,0 +1,441 @@
+package auth
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/bluesky-social/indigo/atproto/identity"
+	jose "github.com/go-jose/go-jose/v3"
+	"github.com/go-jose/go-jose/v3/jwt"
+	"github.com/rs/zerolog/log"
+	"golang.org/x/oauth2"
+)
+
+type ClientMetadata struct {
+	ClientName              string              `json:"client_name"`
+	ClientId                string              `json:"client_id"`
+	ClientUri               string              `json:"client_uri"`
+	ApplicationType         string              `json:"application_type"`
+	GrantTypes              []string            `json:"grant_types"`
+	Scope                   string              `json:"scope"`
+	ResponseTypes           []string            `json:"response_types"`
+	RedirectUris            []string            `json:"redirect_uris"`
+	TokenEndpointAuthMethod string              `json:"token_endpoint_auth_method"`
+	TokenEndpointAuthSigner string              `json:"token_endpoint_auth_signing_alg"`
+	DpopBoundAccessTokens   bool                `json:"dpop_bound_access_tokens"`
+	Jwks                    *jose.JSONWebKeySet `json:"jwks"`
+}
+
+type OAuthClient interface {
+	ClientMetadata() *ClientMetadata
+	Authorize(dpopClient *DpopHttpClient, i *identity.Identity, handle *string) (string, *AuthorizeState, error)
```

Comment:
any reason `handle` needs to be a pointer?

Reply:
It's misnamed, should really be `loginHint` which would be the same as a handle, but it should still function if we don't know the handle beforehand.